### PR TITLE
feat: add .luarc.json

### PR DESCRIPTION
**Pull request by lewis6991**
Update and promote `contrib/luarc.json` to `.luarc.json`.

- Nvim's LSP settings are now maintained.
- This will make diagnostics consistent across users.
- Will be required for #24563

---

### Adapting LSP configuration

Use something like the following so RTP isn't added to projects with a `.luarc.json`.

<details>
<summary>Via nvim-lspconfig</summary>

```lua
local nvim_lsp = require('lspconfig')

nvim_lsp.lua_ls.setup {
  on_init = function(client)
    local path = client.workspace_folders[1].name
    if not vim.uv.fs_stat(path..'/.luarc.json') then
      -- Make the server aware of Neovim runtime files
      client.config.settings.Lua.workspace.library = { vim.env.VIMRUNTIME }
      -- or for everything:
      -- client.config.settings.Lua.workspace.library = vim.api.nvim_get_runtime_file("", true)
      client.notify("workspace/didChangeConfiguration", {
        settings = client.config.settings
      })
    end
  end
}
```
</details>

<details>
<summary>Manually</summary>

```lua
local function find_root(path, markers)
  local match = vim.fs.find(markers, { path = path, upward = true })[1]
  if not match then
    return
  end
  local stat = vim.loop.fs_stat(match)
  local isdir = stat and stat.type == "directory"
  return vim.fn.fnamemodify(match, isdir and ':p:h:h' or ':p:h')
end

vim.api.nvim_create_autocmd('FileType', {
  pattern = 'lua',
  group = vim.api.nvim_create_autocmd('luals', {}),
  callback = function(args)
    vim.lsp.start({
      name = 'luals',
      filetype = 'lua',
      cmd = { 'lua-language-server' },
      root_dir = find_root(args.file, {'.luarc.json', '.git'}),
      on_init = function(client)
        local path = client.workspace_folders[1].name
        if not vim.uv.fs_stat(path..'/.luarc.json') then
          -- Make the server aware of Neovim runtime files
          client.config.settings.Lua.workspace.library = { vim.env.VIMRUNTIME }
          -- or for everything:
          -- client.config.settings.Lua.workspace.library = vim.api.nvim_get_runtime_file("", true)
          client.notify("workspace/didChangeConfiguration", {
            settings = client.config.settings
          })
        end
      end,
    })
  end,
})
```
</details>

### CODE REVIEW

File: .luarc.json [hunk 1]
```diff
@@ -0,0 +1,29 @@
+{
+  "$schema": "https://raw.githubusercontent.com/sumneko/vscode-lua/master/setting/schema.json",
```
Comment (justinmk) [inline .luarc.json:2]:
```suggestion
  "$schema": "https://raw.githubusercontent.com/LuaLS/vscode-lua/master/setting/schema.json",
```

**Review by justinmk — APPROVED** [review]

**Comment by gpanders** [comment]
I'm not opposed to this, my _only_ thought is that if I want to modify `.luarc.json` to add e.g. `neodev.nvim` to the library paths, is there a way to do that without my worktree constantly being dirty?

**Comment by lewis6991** [comment]
There's no need for neodev anymore.

Or is there? I've added meta files for everything now. Anything that's missing should be added as PR's here instead of relying on third party code.

**Comment by gpanders** [comment]
> There's no need for neodev anymore.
> 
> Or is there?
> 
> I've added meta files for everything now. Anything that's missing should be added as PR's here (via options.lua or eval.lua) instead of relying on third party code.

Good point, you're probably right.

**Review by gpanders — APPROVED** [review]

File: .luarc.json [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    "groupSeverity": {
+      "strong": "Warning",
+      "strict": "Warning"
+    },
+    "unusedLocalExclude": [ "_*" ],
+    "disable": [
+      "luadoc-miss-see-name"
+    ]
+  },
+  "telemetry.enable": false
```
Comment (clason) [inline .luarc.json:28]:
```suggestion
```
That's been removed a long time ago.

**Comment by MariaSolOs** [comment]
> There's no need for neodev anymore.

@lewis6991 Could you please elaborate on this? I actually [struggled recently with duplicate results when using `neodev`](https://www.reddit.com/r/neovim/comments/15kv7dv/duplicate_definitionshover_with_neodev/), but when disabling I was unsure of how exactly to configure `lua_ls` in order to get the correct results. Should [these default settings](https://github.com/neovim/nvim-lspconfig/blob/b6091272422bb0fbd729f7f5d17a56d37499c54f/doc/server_configurations.md#lua_ls) from `nvim-lspconfig` be enough?

**Comment by clason** [comment]
I too would like to know how this should be used in combination with the default config for, say, working on plugins or init.lua. Right know, luals loads everything multiple times...

**Comment by lewis6991** [comment]
neodev should be disabled when working on the Neovim codebase. The neodev README explains how to do this.

**Comment by clason** [comment]
I'm not using neodev, though; just manual config.

**Comment by MariaSolOs** [comment]
@lewis6991 by disabling neodev in this codebase you mean using neoconf? I didn’t find anything else mentioning that in the README, but I might have missed it.

So I’m now confused about whether I need neodev at all, but I’m still unsure of what modifications I need to do to my lua_ls config :/

**Comment by clason** [comment]
Snippet?

**Comment by MariaSolOs** [comment]
> Don't include:
> 
> ```lua
>     workspace = {
>         -- Make the server aware of Neovim runtime files
>         library = vim.api.nvim_get_runtime_file("", true),
>       }
> ```
> 
> When attaching to core.

Pardon my ignorance, but wdym by attaching to core?

**Comment by lewis6991** [comment]
> @lewis6991 by disabling neodev in this codebase you mean using neoconf? 

No

> So I’m now confused about whether I need neodev at all, but I’m still unsure of what modifications I need to do to my lua_ls config :/

You don't need it. Just don't set library in your settings.

**Comment by lewis6991** [comment]
> Pardon my ignorance, but wdym by attaching to core? 

This repository.

**Comment by lewis6991** [comment]
> Snippet?
> 

Maybe look at my config. https://github.com/lewis6991/dotfiles/blob/main/config/nvim/lua/lewis6991/lsp.lua

Basically, just don't set library.

**Comment by clason** [comment]
Well, but if I _need_ `library` when working on, say, nvim-treesitter?

**Comment by lewis6991** [comment]
> Well, but if I _need_ `library` when working on, say, nvim-treesitter? 

Either add a .luarc.json to nvim-treesitter (like I do). Or dynamically setup library based on the root directory.

**Comment by clason** [comment]
So you just made many people's lives more complicated. Color me unimpressed.

**Comment by MariaSolOs** [comment]
> > Pardon my ignorance, but wdym by attaching to core?
> 
> This repository.

Does this mean that I need to build neovim from source? I’m confused lol. As someone that just wants to get completions for `vim` APIs when tweaking my `.config/nvim` files, it sounds like I can just ditch neodev, make no further changes to my LSP configuration, and it _should just work_?

Sorry I’m slow. I’m not very smart at 1am. 

Anyway I think we should probably talk to Folke when he’s back from vacation because it would be a good idea to make these clarifications in neodev.

**Comment by MariaSolOs** [comment]
> > As someone that just wants to get completions for vim APIs when tweaking my .config/nvim files
> 
> In this case, this PR is completely irrelevant, and nothing has changed for you. (The file is only used when editing files in a cloned `neovim` repo.)

Oh lol 😂 well sorry then for the spam

**Comment by lewis6991** [comment]
Added some Lua snippets for LSP configuration to the description.

**Comment by lewis6991** [comment]
Raised: https://github.com/neovim/nvim-lspconfig/pull/2750

**Comment by towry** [comment]
@MariaSolOs Seems neodev plugin is not needed.

I have created a `.luarc.json` file in my neovim config folder with following content:

```json
{
  "$schema": "https://raw.githubusercontent.com/sumneko/vscode-lua/master/setting/schema.json",
  "workspace": {
    "library": [
      "/usr/local/share/nvim/runtime/lua",
      "/Users/towry/.local/share/nvim/lazy/nvim-cmp/lua"
    ]
  }
}
```

Here are the effects:

截屏2023-08-08 21 12 53

截屏2023-08-08 21 07 50

截屏2023-08-08 21 06 10


---

- https://github.com/LuaLS/lua-language-server/wiki/Libraries#configuration-file

**Comment by lewis6991** [comment]
You can replace `/usr/local/share/nvim/runtime/lua` with `$VIMRUNTIME`.

**Comment by MariaSolOs** [comment]
@towry @lewis6991 thanks a lot for the clarifications and helpful snippets <3

> In this case, this PR is completely irrelevant, and nothing has changed for you. (The file is only used when editing files in a cloned neovim repo.)

Also FWIW I do have a `neovim` clone and I do want to become an active contributor, so this is still relevant to me :)

**Comment by MariaSolOs** [comment]
@lewis6991 why does your snippet for `on_init` return `true`? From what I can see [here](https://github.com/neovim/neovim/blob/8afdc1f3867a620c8235b3d3964b019b94657190/runtime/lua/vim/lsp.lua#L1437), the return value is ignored on success.

Sorry if it's a silly question, just making sure I'm not missing something.

**Comment by MariaSolOs** [comment]
> You can replace `/usr/local/share/nvim/runtime/lua` with `$VIMRUNTIME` and the other line can maybe use `$XDG_DATA_HOME`.

not sure if this is a local issue but I just couldn't make the paths to plugins work using `$XDG_DATA_HOME`. The only way I could make `lua_ls` pick up the types was by hardcoding the path :/

**Comment by lewis6991** [comment]
I think you are right.

I used https://github.com/neovim/nvim-lspconfig/wiki/Project-local-settings as a template, but that must be outdated.

**Comment by neovim-discourse** [comment]
This pull request has been mentioned on **Neovim Discourse**. There might be relevant details there:

https://neovim.discourse.group/t/how-to-configure-lsp-for-editing-init-lua-to-aid-in-setting-up-neovim/4269/1

**Pull request merged**